sdk/nodejs/runtime/rpc.ts

```diff
@@ -578,54 +586,113 @@ export function suppressUnhandledGrpcRejections<T>(p: Promise<T>): Promise<T> {
     return p;
 }
 
+function sameVersion(a?: semver.SemVer, b?: semver.SemVer): boolean {
+    // We treat undefined as a wildcard, so it always equals every other version.
+    return a === undefined || b === undefined || semver.eq(a, b);
+}
+
+function checkVersion(want?: semver.SemVer, have?: semver.SemVer): boolean {
+    if (want === undefined || have === undefined) {
+        return true;
+    }
+    return have.major === want.major && have.minor >= want.minor && have.patch >= want.patch;
+}
+
 /**
  * A ResourcePackage is a type that understands how to construct resource providers given a name, type, args, and URN.
  */
 export interface ResourcePackage {
-    constructProvider(name: string, type: string, args: any, opts: { urn: string }): ProviderResource;
+    readonly version: semver.SemVer | undefined;
+    constructProvider(name: string, type: string, urn: string): ProviderResource;
 }
 
-const resourcePackages = new Map<string, ResourcePackage>();
-
-function packageKey(name: string, version: string): string {
-    return `${name}@${version}`;
-}
+const resourcePackages = new Map<string, ResourcePackage[]>();
 
 /**
  * registerResourcePackage registers a resource package that will be used to construct providers for any URNs matching
  * the package name and version that are deserialized by the current instance of the Pulumi JavaScript SDK.
  */
-export function registerResourcePackage(name: string, version: string, pkg: ResourcePackage) {
-    const key = packageKey(name, version);
-    const existing = resourcePackages.get(key);
-    if (existing) {
-        throw new Error(`Cannot re-register package ${key}. Previous registration was ${existing}, new registration was ${pkg}.`);
+export function registerResourcePackage(pkg: string, resourcePackage: ResourcePackage) {
+    let packages = resourcePackages.get(pkg);
+    if (packages) {
+        for (const existing of packages) {
+            if (sameVersion(existing.version, resourcePackage.version)) {
+                throw new Error(`Cannot re-register package ${pkg}@${resourcePackage.version}. Previous registration was ${existing}, new registration was ${resourcePackage}.`);
+            }
+        }
+    } else {
+        packages = [];
+        resourcePackages.set(pkg, packages);
     }
-    resourcePackages.set(key, pkg);
+
+    log.debug(`registering package ${pkg}@${resourcePackage.version}`);
+    packages.push(resourcePackage);
+}
+
+export function getResourcePackage(pkg: string, version: string): ResourcePackage | undefined {
+    const ver = semver.parse(version) ?? undefined;
+
+    let bestPackage: ResourcePackage | undefined = undefined;
+    for (const resourcePackage of resourcePackages.get(pkg) ?? []) {
+        if (!checkVersion(ver, resourcePackage.version)) {
+            continue;
+        }
+        if (bestPackage === undefined ||
+            (resourcePackage.version && bestPackage.version && semver.gt(resourcePackage.version, bestPackage.version))) {
```

Comment:
Is this correct?

Reply:
I think that this should use `checkVersion`, which accommodates the `undefined` wildcard behavior.

Reply:
Likewise for modules.